File: sheet/prog/regex.py
#!/usr/bin/env python3

import re

# r""で、\をそのまま使用できる
pattern = r"([\w+\.]+)@(\w+\.[\w]+)"
content = "kcrt@kcrt.net"

# POSIXキャラクタクラス([:alpha:]など)は使用できない

ret = re.match(pattern, content)
if ret:
    print(ret.group(0))  # kcrt@kcrt.net
    print(ret.group(1))  # kcrt
else:
    print("No match!")


# 名前付きキャプチャ: (?P<name>...)
ret = re.match(r"(?P<hour>\d\d):(?P<min>\d\d)", "12:34")
ret.group("hour")   # => 12

# オプションは
re.match(pattern, content, re.IGNORECASE)
#  re.ASCII (re.A)      ASCIIモード(ないとアラビア数字以外も\dでマッチ)
#  re.LOCALE (re.L)     ローカル言語モード (非推奨)
#  re.UNICODE (re.U)    Unicodeモード (指定不要)
#  re.IGNORECASE (re.I) 大文字小文字無視
#  re.MULTILINE (re.M)  複数行
#  re.DEBUG, re.VERBOSE(re.X)

# 事前にコンパイルしておくことで再利用･高速化できる
# 上記オプションも指定可能
pat = re.compile(pattern, re.ASCII)
pat.match(content)

# 全体一致でなくても良い時
re.search(pattern, content)

# マッチした物をすべて探す(リスト・イテレータ)
re.findall(pattern, content)
re.finditer(pattern, content)


# 置き換え
re.sub("world", "xxxxx", "Hello, world.")
# => Hello, xxxxx
re.sub("world", r"*\g<0>*", "Hello, world.")
# => Hello, *world*
# 対応グループは\1 (== \g<1>)という表記もできる
# \0は使用できないので\g<0>を使用すること

File: sheet/prog/regex.jl
#!/usr/bin/env julia

# r""はregex型
pattern = r"([[:word:]+\.]+)@([[:word:]]+\.[[:word:]]+)"
content = "kcrt@kcrt.net"
typeof(pattern)		# => Regex

# POSIXキャラクタクラス([:alpha:]など)は使用できる

occursin(pattern, content)	# => true
ret = match(pattern, content)	# => RegexMatchを返す
if ret !== nothing
	println(ret.match)
	println(ret.captures[1])
else
	println("No match!")
end

# 名前付きキャプチャ: (?P<name>...)
ret = match(r"(?P<hour>\d\d):(?P<min>\d\d)", "12:34")
ret[:hour]	# => 12

# r文字列はオプションも指定できる
occursin(r"apple"i, "APPLE")
# i: 大文字小文字無視, m: 複数行, s: 単行, x: スペースを無視


# 文字列からRegexを作成する場合は
domain = "kcrt.net"
pat = Regex(".*@$domain")

# 置き換え replace(str, r"pattern1" => s"pattern2")
# s""はSubstituionString型
replace("Hello, world", "world" => "xxxxx")
# => Hello, xxxxx
replace("Hello, world", r"w...d" => s"*\g<0>*")
# => Hello, *world*
